packages/ui/src/lib/Tooltip.svelte

```diff
@@ -0,0 +1,142 @@
+<script lang="ts" context="module">
+	export type TooltipPosition = 'top' | 'bottom';
+	export type TooltipAlign = 'start' | 'center' | 'end';
+</script>
+
+<script lang="ts">
+	import { portal } from '$lib/utils/portal';
+	import { flyScale } from '$lib/utils/transitions';
+	import { type Snippet } from 'svelte';
+
+	interface Props {
+		text?: string;
+		delay?: number;
+		align?: TooltipAlign;
+		position?: TooltipPosition;
+		children: Snippet;
+	}
+
+	const { text, delay = 700, align = 'center', position = 'bottom', children }: Props = $props();
+
+	let targetEl: HTMLElement | undefined = $state();
+	let tooltipEl: HTMLElement | undefined = $state();
+
+	let show = $state(false);
+	let timeoutId: undefined | ReturnType<typeof setTimeout> = $state();
+
+	const isTextEmpty = $derived(!text || text === '');
+
+	function handleMouseEnter() {
+		timeoutId = setTimeout(() => {
+			show = true;
+			// console.log('showing tooltip');
+		}, delay); // 500ms delay before showing the tooltip
+	}
+
+	function handleMouseLeave() {
+		clearTimeout(timeoutId);
+		show = false;
+	}
+
+	function adjustPosition() {
+		if (!targetEl || !tooltipEl) return;
+
+		const tooltipRect = tooltipEl.getBoundingClientRect();
+		// get first child of targetEl
+		const targetChild = targetEl.children[0];
+		const targetRect = targetChild.getBoundingClientRect();
+
+		let top = 0;
+		let left = 0;
+		let transformOriginTop = 'center';
+		let transformOriginLeft = 'center';
+		const gap = 4;
+
+		if (position === 'bottom') {
+			top = targetRect.bottom + window.scrollY + gap;
+
+			transformOriginTop = 'top';
+		} else if (position === 'top') {
+			top = targetRect.top - tooltipRect.height + window.scrollY - gap;
+
+			transformOriginTop = 'bottom';
+		}
+
+		if (align === 'start') {
+			left = targetRect.left + window.scrollX;
+			transformOriginLeft = 'left';
+		} else if (align === 'end') {
+			left = targetRect.right - tooltipRect.width + window.scrollX;
+			transformOriginLeft = 'right';
+		} else if (align === 'center') {
+			left = targetRect.left + targetRect.width / 2 - tooltipRect.width / 2 + window.scrollX;
+			transformOriginLeft = 'center';
+		}
+
+		tooltipEl.style.top = `${top}px`;
+		tooltipEl.style.left = `${left}px`;
+		tooltipEl.style.transformOrigin = `${transformOriginTop} ${transformOriginLeft}`;
+	}
+
+	$effect(() => {
+		if (tooltipEl) {
+			adjustPosition();
+		}
+	});
+</script>
+
+{#if isTextEmpty}
+	{@render children()}
+{:else}
+	<span
+		bind:this={targetEl}
+		class="tooltip-wrap"
+		role="tooltip"
+		onmouseenter={handleMouseEnter}
+		onmouseleave={handleMouseLeave}
+	>
+		{#if children}
+			{@render children()}
+		{/if}
+
+		{#if show}
+			<div
+				bind:this={tooltipEl}
+				use:portal={'body'}
+				class="tooltip-container text-11 text-body"
+				transition:flyScale={{
+					position: position
+				}}
+			>
+				<span>{text}</span>
+			</div>
+		{/if}
+	</span>
```

Comment:
In the past we had a tooltip component that surrounded the target with an element this way, but I felt like it introduced problems in various circumstances. Do you feel confident this works e.g. when you have an element inside an `overflow: none` container?

Reply:
It shouldn’t be a problem. I also don’t like it when components like this are rendered inside their targets or triggers. The component will be rendered outside the element in the ‘body’ using the ‘portal’ hook.

Reply:
I didn't provide this in the PR description: the benefit with this approach is to use svelte components and transitions + isolate tooltip styles.

Reply:
A'ight, but let me know if you run into any problems in the future. I personally don't like the wrapping container for anything with a tooltip, but it's a moot point if there aren't any negative consequences.

The alternative solution I could think of is using a directive still, but having single component responsible for rendering, like ToastController.

Reply:
Thanks! Let's see how it goes. There will be another similar component — a mini-card tooltip that appears on hover— where I should be able to pass Svelte components like links, images, etc. This change is an experiment to see if we can create a tooltip as a Svelte component.

<img width="419" alt="image" src="https://github.com/user-attachments/assets/dc8b6f38-11eb-49ca-bac0-cd9eda143742">